lib/core/src/Cardano/Wallet/Network.hs

```diff
@@ -154,35 +162,60 @@ defaultRetryPolicy =
 
 -- | Subscribe to a blockchain and get called with new block (in order)!
 follow
-    :: NetworkLayer IO tx block
+    :: Show e
+    => NetworkLayer IO tx block
     -- ^ The @NetworkLayer@ used to poll for new blocks.
     -> Trace IO Text
     -> BlockHeader
     -- ^ The local tip to start at. Blocks /after/ the tip will be yielded.
     -> (NE.NonEmpty block -> BlockHeader -> ExceptT e IO ())
-     -- ^ Callback with blocks and the current tip of the /node/. @follow@ stops
-     -- polling and terminates if the callback errors.
+    -- ^ Callback with blocks and the current tip of the /node/. @follow@ stops
+    -- polling and terminates if the callback errors.
     -> (block -> BlockHeader)
+    -- ^ Getter on the abstract 'block' type
     -> IO ()
-follow nl tr start yield header = do
-    liftIO $ runExceptT (networkTip nl) >>= \case
-        Right nwTip ->
-            restoreStep (start, nwTip)
-        Left e -> do
-            logError tr $ "Failed to get network tip: " +|| e ||+ ""
-            restoreSleep start
+follow nl tr start yield header =
+    sleep 0 start
   where
-    restoreStep
-        :: (BlockHeader, BlockHeader)
-        -> IO ()
-    restoreStep (localTip, nodeTip) = do
+    pause :: Int
+    pause = 2*1000*1000 -- 2 seconds
+
+    -- | Wait a short delay before querying for blocks again. We also take this
+    -- opportunity to refresh the chain tip as it has probably increased in
+    -- order to refine our syncing status.
+    sleep :: Int -> BlockHeader -> IO ()
+    sleep delay localTip = do
+        when (delay > 0) (threadDelay delay)
+        let io = runExceptT (networkTip nl) >>= \case
+                Right nodeTip ->
+                    step (localTip, nodeTip)
+                Left e -> do
+                    logWarning tr $ T.pack $
+                        "Failed to get network tip: " <> show e
+                    sleep delay localTip
+        recovering policy [shouldRetry] (const io)
```

Comment:
It feels a bit nuanced in combination with the existing stepping. It seems we will recursively create new `recovering` too? Would it not be simpler code-wise and better for users if we used the existing `2 s` sleep? (or at the least settled for one solution)

If not, can we at least add a comment before line 196 reminding that `recovering` will retry IO exceptions and not `Left e`.

Also relaunching the node with a different chain during syncing resulted in these warnings for me, but thought it would be errors?
```
[iohk.cardano-wallet.serve.worker.54b5c2a8:Warning:ThreadId 38] [2019-10-04 10:57:33.95 UTC] Failed to get next blocks: ErrGetBlockNotFound (Hash {getHash = "t\164|QU\145\184\217}\237^\ETB>K\NAK\133\183\197\189\205\204i\GSU\132\&8O\194;\191\165\210"})
```
so I have not managed to see this code in action.

Reply:
Ah, right, never mind about the error, it is expected of course
https://github.com/input-output-hk/cardano-wallet/blob/b6f118b4243cc8e1bbd373c79d4cb7b6aa6262ee/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs#L407